Open XL Toolchain Changes

Deigue opened the pull request:
Contains all the initial required toolchain changes to start using Open XL compiler with z/OS builds.

---

Keeping as draft state until I have verified it does not interfere with z/OS XLC Builds.

Deigue commented:
@joransiu @keithc-ca Created draft PR for toolchain changes for some early feedback. Let me know if I should add/tag anyone who might recommend good practice changes that might be needed/missing from these.

keithc-ca commented on runtime/cmake/platform/os/zos.cmake:
```diff
@@ -25,7 +25,10 @@ set(CMAKE_SHARED_LIBRARY_PREFIX "lib")
-    "\"-Wc,list(),offset,gonumber\""
-)
+# todo: perhaps this belongs in corresponding toolchain file?
+if($(TOOLCHAIN_TYPE),xlc)
```
Variables are referenced using `${var}`, the toolchain is in `OMR_TOOLCONFIG` and this comparison needs to use `STREQUAL`:
```
if(${OMR_TOOLCONFIG} STREQUAL "xlc")
```

keithc-ca commented on runtime/cmake/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,59 @@
+################################################################################
+# Copyright IBM Corp. and others 2020
```
As this is a new file, the year should be 2024.

keithc-ca commented on runtime/cmake/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,59 @@
+	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-exceptions -g")
+	list(APPEND OMR_PLATFORM_COMPILE_OPTIONS -fno-rtti)
+else()
+	# xlc/xlc++ options
```
This file should only have OpenXL-style options, not those for xlc/xlc++.

keithc-ca commented on runtime/compiler/makefile.ftl:
```diff
@@ -54,8 +54,7 @@ export BUILD_CONFIG?=prod
 <#if uma.spec.id?starts_with("zos_390-64")>
-  export PLATFORM=s390-zos64-vacpp
-  export ASMLIST=1
+  export PLATFORM=s390-zos64-openxl
```
I think we should focus (at least at first) on the cmake changes necessary to support OpenXL. This change and all the `.mk` file changes should be removed (deferred) until we find a need to support UMA-style builds for z/OS (in Java 11+).

keithc-ca reviewed (changes requested):
Please remove all changes that aren't related to use of cmake.

Deigue commented:
I will undo the UMA/mk updates (they were some ported commits of some much earlier attempts at open xl compilation) and verify that it doesn't impact the current compilation with java 21, along with the above suggestions.
Thanks for the feedback.

pshipton commented:
We'll also have to verify IBM Java 8 compilation isn't impacted.

Deigue commented on runtime/cmake/platform/os/zos.cmake:
```diff
@@ -25,7 +25,10 @@ set(CMAKE_SHARED_LIBRARY_PREFIX "lib")
+if($(TOOLCHAIN_TYPE),xlc)
+    list(APPEND OMR_PLATFORM_COMPILE_OPTIONS
+        "\"-Wc,inline(auto,noreport,600,5000)\""
+        "\"-Wc,list(),offset,gonumber\""
```
https://github.ibm.com/runtimes/openj9-openjdk-jdk21-zos/issues/320
Opened an issue as this might be a more extended discussion and is not directly supported. We might need to communicate with the Open XL team the necessity of these, or if there are equivalent alternatives.

Deigue commented:
Removed all changes that aren't related to use of cmake and addressed the comments above or added comments as necessary.
I have also verified now that this maintains the compilation progress so-far for Open XL.

Pending verification of a standard xlc build and a java 8 build to ensure no impact.

keithc-ca commented on runtime/compiler/compiler.mk:
```diff
@@ -53,6 +53,17 @@ ifeq ($(PLATFORM),ppc64-linux64-clang)
     endif
 endif
 
+ifeq ($(PLATFORM),s390-zos64-openxl)
+    # todo: are these hardcoded default paths needed?
+    ifeq (default,$(origin CC))
+        export CC=/jit/team/tsjit/openxl/usr/lpp/IBM/cnw/v2r1/openxl/bin/ibm-clang64
+    endif
+    ifeq (default,$(origin CXX))
+        export CXX=/jit/team/tsjit/openxl/usr/lpp/IBM/cnw/v2r1/openxl/bin/ibm-clang64++
+    endif
+    export A2E_INCLUDE_PATH?=/usr/.lpp/cbclib/include
+endif
```
Please revert changes to this file: it is not used in cmake builds.

Deigue replied:
sorry about that, seems I missed this one.

keithc-ca commented on runtime/cmake/platform/os/zos.cmake:
```diff
@@ -25,7 +25,9 @@ set(CMAKE_SHARED_LIBRARY_PREFIX "lib")
-    "\"-Wc,list(),offset,gonumber\""
-)
+if(${OMR_TOOLCONFIG} STREQUAL "xlc")
+    list(APPEND OMR_PLATFORM_COMPILE_OPTIONS
+        "\"-Wc,inline(auto,noreport,600,5000)\""
```
Please indent with tabs like (most) other cmake files.

Deigue commented:
Removed `compiler.mk` and indented the changes in `runtime/cmake/platform/os/zos.cmake` appropriately.

Pending build verification

keithc-ca commented on runtime/cmake/platform/os/zos.cmake:
```diff
@@ -25,7 +25,9 @@ set(CMAKE_SHARED_LIBRARY_PREFIX "lib")
+		"\"-Wc,inline(auto,noreport,600,5000)\""
+		"\"-Wc,list(),offset,gonumber\""
+	)
+endif()
```
Perhaps these options belong in `xlc.cmake`?
Are there equivalent options for OpenXL?

Deigue replied:
As mentioned above with Peters intial comment, a issue has been created: https://github.ibm.com/runtimes/openj9-openjdk-jdk21-zos/issues/320 ...
There are no current exact equivalents, and the cpp team has suggested to not put equivalent optimization flags as there is no guarantee it will benefit or worsen the performance.

I can move this to `xlc.cmake` though if thats better ...

Deigue replied:
Moved into `xlc.cmake` in https://github.com/eclipse-openj9/openj9/commit/1be349dbb7cbfe8c1a429e852e0b71f08b35489f

keithc-ca commented on runtime/cmake/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,35 @@
+endif()
+
+# OMR_PLATFORM_CXX_COMPILE_OPTIONS gets applied to the jit (which needs exceptions),
+# so we put these in the CMAKE_CXX_FLAGS instead
```
Please finish each sentence with a period.

Deigue replied:
Done.

keithc-ca commented on runtime/cmake/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,35 @@
+
+# OMR_PLATFORM_CXX_COMPILE_OPTIONS gets applied to the jit (which needs exceptions),
+# so we put these in the CMAKE_CXX_FLAGS instead
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fignore-exceptions")
```
This doesn't seem right. JIT uses (throws and catches) exceptions: how does ignoring them make sense?

Deigue replied:
I was trying to recall why I put this here, I think I have found it.
It is the equivalence for this here in xlc.cmake: https://github.com/eclipse-openj9/openj9/blob/2a2df9f1117f0e3dad2ab9022dc6d05734c0cf83/runtime/cmake/platform/toolcfg/xlc.cmake#L35

Mapping of that option goes to `-fignore-exceptions` as per https://www.ibm.com/docs/en/open-xl-c-cpp-zos/2.1?topic=options-mapping

keithc-ca replied:
I misread the reference in the comment to JIT code: this option is only for non-JIT code.

Deigue commented:
Comments above addressed ... 
Launched builds [j21](https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK21_s390x_zos_Personal/757/console) and [java 8](http://vmfarm.rtp.raleigh.ibm.com/build_info.php?build_id=73556) to verify. 

Once those check out, will bring PR out of draft.

keithc-ca commented on runtime/cmake/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,35 @@
+# SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+################################################################################
+
+list(APPEND OMR_PLATFORM_COMPILE_OPTIONS -O3)
+
+list(APPEND OMR_PLATFORM_CXX_COMPILE_OPTIONS -fno-rtti)
+
+list(APPEND OMR_PLATFORM_COMPILE_OPTIONS -fstack-protector)
```
Please combine the two contributions to `OMR_PLATFORM_COMPILE_OPTIONS`:
```
list(APPEND OMR_PLATFORM_COMPILE_OPTIONS
	-O3
	-fstack-protector
)
```

keithc-ca commented on runtime/cmake/platform/toolcfg/xlc.cmake:
```diff
@@ -20,7 +20,11 @@
 ################################################################################
 
-list(APPEND OMR_PLATFORM_COMPILE_OPTIONS -O3)
+list(APPEND OMR_PLATFORM_COMPILE_OPTIONS
+		-O3
+		"\"-Wc,inline(auto,noreport,600,5000)\""
+		"\"-Wc,list(),offset,gonumber\""
+	)
```
Please format this like other places that append multiple options:
```
list(APPEND OMR_PLATFORM_COMPILE_OPTIONS
	-O3
	"\"-Wc,inline(auto,noreport,600,5000)\""
	"\"-Wc,list(),offset,gonumber\""
)
```

Deigue commented:
Thanks, I've addressed the above comments.

Deigue commented:
@keithc-ca This one is ready for a re-check.

keithc-ca commented on runtime/cmake/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,36 @@
+list(APPEND OMR_PLATFORM_CXX_COMPILE_OPTIONS -fno-rtti)
+
+if(NOT OMR_OS_ZOS)
+	list(APPEND OMR_PLATFORM_CXX_COMPILE_OPTIONS -qsuppress=1540-1087:1540-1088:1540-1090)
```
I don't think Open XL supports any options beginning with `-q`; this looks like stuff carried over from xlc.

Deigue replied:
Done, removed the unused options, and also rebased with latest commit.

Deigue commented:
Rebased with master, confirmed passing with java 8 and java 21 builds.

keithc-ca commented on runtime/cmake/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,32 @@
+
+list(APPEND OMR_PLATFORM_CXX_COMPILE_OPTIONS -fno-rtti)
+
+# OMR_PLATFORM_CXX_COMPILE_OPTIONS gets applied to the jit (which needs exceptions),
```
Suggest this comment could be improved:
```
# OMR_PLATFORM_CXX_COMPILE_OPTIONS get applied to JIT code (which needs exceptions),
```

Deigue replied:
Updated comment

keithc-ca commented:
Jenkins compile aix jdk21
